### DISCUSSION TITLE
詳しい動かし方について

### DISCUSSION THREAD

**Discussion by secretnmae** [tetsuya-ki/discord-live-notificationbot#21] [Q&A]
初めまして。
Discordのツイキャス用の配信告知BOTを探していて唯一こちらを見つけました。
しかしながら、知識不足のため、動かし方が全く分かりません。
恐縮ですが、詳しい動かし方について一から教えていただけると幸いです。
よろしくお願いいたします。

**Comment by tetsuya-ki** [comment]
使い方については下記に書いてあります。
https://github.com/tetsuya-ki/discord-live-notificationbot/#live-notification_add
Botを自分のサーバーに追加し、`/live`を入力するとBotを操作できます
![image](https://user-images.githubusercontent.com/64536338/192140568-f122e117-55a3-4205-9990-8cbc8c48617a.png)

「動かし方」というのは自分でBotを動かすということでしょうか？
(1)はい＞自分で動かす
動かす環境はWindows、Mac、Linuxどれかで地味に変わると思います。WIndowsはWSL2を使えばLinuxと同様にできます
1. [Pythonをインストール](https://www.python.jp/install/install.html#4jqgmS)
2. [Poetryをインストール](https://cocoatomo.github.io/poetry-ja/)
3. envファイル(環境変数)を準備
[.env.sample](https://github.com/tetsuya-ki/discord-live-notificationbot/blob/main/cogs/modules/files/.env.sample)をコピーし、同じディレクトリに`.env`ファイルを配置する
[環境変数](https://github.com/tetsuya-ki/discord-live-notificationbot#%E7%92%B0%E5%A2%83%E5%A4%89%E6%95%B0)の説明を確認し、少なくとも以下2つは必須
  - DISCORD_TOKEN
  - GUILD_ID_FOR_ATTACHMENTS

4. [動かす](https://github.com/tetsuya-ki/discord-live-notificationbot/#%E5%8B%95%E3%81%8B%E3%81%99)

(2)いいえ＞自分で動かさない
自分で動かさない場合、現時点では無理です。サンプルとしてBotを公開していましたが、100サーバーに登録されているため、これ以上追加できません。
また、上記で動かしているBotも[YouTubeから情報取得できなくなる事象](https://github.com/tetsuya-ki/discord-live-notificationbot/issues/19)や、replitのため頻繁に止まるなどあるので、期待した品質に達していないと思います。

**Comment by secretnmae** [comment]
ご丁寧に返信いただきありがとうございます。
二つほど質問してもよろしいでしょうか。
・ギルドIDとはチャンネルIDのことですか。
・poetry install 実行時にエラーがいくつか出てるので、私自身で調べて対処できないときまた質問してもよろしいですか。

**Reply by tetsuya-ki** [reply]
ギルドIDとは、サーバーIDのことです。設定で開発者モードを有効にすると、サーバー名を右クリック(タップ)してギルドIDをコピーできます。
<https://camon.tokyo/discord/development-bot-id/>

問題ありませんが、質問する際は実行環境について必ず記載してください(OS名、Pythonの環境)。

**Reply by secretnmae** [reply]
お久しぶりです。
最近忙しくてできてなかったのですが、やはりpoetry install実行時に以下のようなエラーが発生し、python.exe -m pip install --upgrade pipを実行し、pipは22.3になっているのですがエラーが解消しません。
(OS:Windows、Python:3.10.6)
![image](https://user-images.githubusercontent.com/114338855/197239906-bb20a094-703e-4fa4-a39f-4e802f67be80.png)
素人質問ではありますが、何卒よろしくお願いいたします。

**Reply by tetsuya-ki** [reply]
Windows11で同じバージョンにしてみたところ、再現しました。
python3.10でやる場合は以下のようにしてください。以下、コマンドプロンプト(管理者権限で起動)してください。
![image](https://user-images.githubusercontent.com/64536338/197335924-5aeba9da-f24e-4aaf-8bb9-c09c0c75d56c.png)

このスレッド？とは別の場所に他のやり方も書いておきます。
```sh
poetry self update
# poetryのバージョンをアップデートします
poetry install
# 依存関係を解決します(今度は失敗しません)
```

```log
# 以下のコマンドで実行できますが、なぜかエラーになりました
poetry run python discord-live-notificationbot.py
=================================
C:\Users\maruk\discord-live-notificationbot>poetry run python discord-live-notificationbot.py
Traceback (most recent call last):
  File "C:\Users\maruk\discord-live-notificationbot\discord-live-notificationbot.py", line 8, in <module>
    import keep_alive
（！！！省略！！！）
  File "C:\Users\maruk\discord-live-notificationbot\.venv\lib\site-packages\jinja2\filters.py", line 13, in <module>
    from markupsafe import soft_unicode
ImportError: cannot import name 'soft_unicode' from 'markupsafe' (C:\Users\maruk\discord-live-notificationbot\.venv\lib\site-packages\markupsafe\__init__.py)
```
markupsafeを2.0.1にすると解決するらしいのですが、そのままではインストールしてくれなかったので、
「poetry.lock」を以下のように修正し、保存します。(2.1.1と書いてあった箇所を2.0.1にしています)

![image](https://user-images.githubusercontent.com/64536338/197336120-9033d328-cad3-4da8-a2e5-8d2d66fe9fed.png)

その後、以下のコマンドを実行します。
```sh
poetry update markupsafe==2.0.1
poetry run python discord-live-notificationbot.py
```

**Reply by secretnmae** [reply]
丁寧に教えて頂き誠にありがとうございます。
おかげさまで何とかpoetryの方は実行でき、BOTも起動しました。
しかし、コマンド入力時に以下のエラーが発生し、対処法がわからないのでご教授の方何卒お願いいたします。

> エラーが発生しました([WinError 32] プロセスはファイルにアクセスできません。別のプロセスが使用中です。: 'C:\Users\Desktop\discord-live-notificationbot-main\cogs\modules\files\live.db')

**Reply by tetsuya-ki** [reply]
@secretnmae .envファイルで未暗号化ファイルを削除するのですが、その処理でエラーが発生しているようです。
ローカルで動かす場合は、暗号化しなくてもまぁ良いと思いますので、`.env`ファイルに以下のように設定してください。
```sh
KEEP_DECRYPTED_FILE=TRUE
```
＊上記左側の文字列が存在し、「=」の左側がTRUEになっている場合はFALSEに書き換え、存在しない場合は追記してください。
samoleファイルをコピペで作った場合はFALSEになっていると思います。

ファイル削除のやり方が確かにいまいちだったなぁと思いました。Linux(WSL2含む)やMacで発生したことがないので優先順位はちょっと低めですが直すかもしれせん。

**Reply by secretnmae** [reply]
BOTが正常に起動しました！
長い間本当にありがとうございました！！
